include/deal.II/base/patterns_tools.h

```diff
@@ -0,0 +1,682 @@
+                      const std::unique_ptr<PatternBase> &p = Convert<T>::to_pattern())
+    {
+      AssertThrow(p->match(s), ExcNoMatch(s, *p));
+      std::istringstream is(s);
+      T value;
+      if (std::is_same<T,unsigned char>::value)
+        {
+          int i;
+          is >> i;
+          value=i;
+        }
+      else
+        is >> value;
+      AssertThrow(!is.fail(), ExcMessage("Failed to convert from \"" + s + "\" to the type \""
+                                         +boost::core::demangle(typeid(T).name()) + "\""));
+      return value;
+    }
+  };
+
+
+  namespace internal
+  {
+    const std::array<std::string, 5> default_list_separator {{" ",  ","  ,  ";"  ,  "|"  ,   "%"}};
+    const std::array<std::string, 5> default_map_separator {{" ",  ":"  ,  "="  ,  "@"  ,   "#"}};
+
+    //specialize a type for all of the STL containers and maps
+    template <typename T>       struct is_stl_container:std::false_type {};
+    template <typename T, std::size_t N> struct is_stl_container<std::array    <T,N>>    :std::true_type {};
+    template <typename... Args> struct is_stl_container<std::vector            <Args...>>:std::true_type {};
+    template <typename... Args> struct is_stl_container<std::deque             <Args...>>:std::true_type {};
+    template <typename... Args> struct is_stl_container<std::list              <Args...>>:std::true_type {};
+    template <typename... Args> struct is_stl_container<std::forward_list      <Args...>>:std::true_type {};
+    template <typename... Args> struct is_stl_container<std::set               <Args...>>:std::true_type {};
+    template <typename... Args> struct is_stl_container<std::multiset          <Args...>>:std::true_type {};
+    template <typename... Args> struct is_stl_container<std::unordered_set     <Args...>>:std::true_type {};
+    template <typename... Args> struct is_stl_container<std::unordered_multiset<Args...>>:std::true_type {};
+    template <typename... Args> struct is_stl_container<std::stack             <Args...>>:std::true_type {};
+    template <typename... Args> struct is_stl_container<std::queue             <Args...>>:std::true_type {};
+    template <typename... Args> struct is_stl_container<std::priority_queue    <Args...>>:std::true_type {};
+
+    template <typename T>       struct is_stl_map:std::false_type {};
+    template <typename... Args> struct is_stl_map<std::map               <Args...>>:std::true_type {};
+    template <typename... Args> struct is_stl_map<std::multimap          <Args...>>:std::true_type {};
+    template <typename... Args> struct is_stl_map<std::unordered_map     <Args...>>:std::true_type {};
+    template <typename... Args> struct is_stl_map<std::unordered_multimap<Args...>>:std::true_type {};
+  }
+
+  //type trait to use the implementation type traits as well as decay the type
+  template <typename T> struct is_stl_container
+  {
+    static constexpr bool const value = internal::is_stl_container<std::decay_t<T>>::value;
```

Comment:
`decay_t` is only available from C++14 on. Using `std::decay<T>::type` instead hopefully doesn't hurt to much :wink: